File: src/python_cibersort/qn.py
"""Bit-exact reimplementation of preprocessCore::normalize.quantiles.

Verified bit-exact against the CRAN Windows binary of preprocessCore
1.72.0 (a) by direct ctypes calls into preprocessCore.dll and (b) by
in-memory comparison inside R (0/333 target rows differ).  Both the
classic (qnorm_c) and NA-handling (qnorm_c_handleNA) entries reduce to
the same arithmetic for NA-free, equal-length input:

1. Target: for each column j (ascending), sort ascending, accumulate
   ``row_mean[i] += sorted[i] / cols`` in plain f64 (divsd/addsd).
2. Distribute: average ranks with tie handling
   (``get_ranks``: tied group [i..k] gets (i+k+2)*0.5), map back to
   original positions; fractional part > 0.4 interpolates as
   ``(target[fl-1] + target[fl]) * 0.5``, else ``target[fl-1]``.

NOTE: comparisons against decimal TSV reference files show ~1-ULP
scatter — that is a decimal round-trip artifact of the file, not the
algorithm; in-memory comparison is exact.

No NA handling (matches qnorm_c, which errors on NA upstream).
"""

from __future__ import annotations

import numpy as np


def quantile_normalize(Y: np.ndarray) -> np.ndarray:
    """normalize.quantiles(Y) — column-wise quantile normalisation."""
    Y = np.asarray(Y, dtype=np.float64)
    rows, cols = Y.shape
    # 1) target distribution, j ascending, divide-then-accumulate (f64)
    row_mean = np.zeros(rows, dtype=np.float64)
    for j in range(cols):
        row_mean += np.sort(Y[:, j]) / cols

    # 2) distribute target back with averaged ranks for ties
    out = np.empty_like(Y)
    for j in range(cols):
        col = Y[:, j]
        order = np.argsort(col, kind="quicksort")  # tie order irrelevant
        sv = col[order]
        ranks = np.empty(rows, dtype=np.float64)
        i = 0
        while i < rows:
            k = i
            while k < rows - 1 and sv[k] == sv[k + 1]:
                k += 1
            if k != i:
                ranks[i : k + 1] = (i + k + 2) * 0.5
            else:
                ranks[i] = i + 1
            i = k + 1
        fl = np.floor(ranks).astype(np.int64)  # floor(rank), 1-based
        frac = ranks - fl
        lo = row_mean[fl - 1]
        hi = row_mean[np.minimum(fl, rows - 1)]  # only used when frac > 0.4
        vals = np.where(frac > 0.4, (lo + hi) * 0.5, lo)
        out[order, j] = vals
    return out

File: rust/src/qn.rs
//! Exact port of preprocessCore::normalize.quantiles (qnorm_c / handleNA path).
//!
//! Verified bit-exact against the CRAN Windows binary of preprocessCore
//! 1.72.0 via direct ctypes calls into preprocessCore.dll and via in-memory
//! comparison inside R (0/333 rows differ): the binary computes
//!   row_mean[i] += sorted_col_j[i] / cols      (f64 divsd+addsd, j ascending)
//! then assigns by averaged ranks; ties with fractional part > 0.4 get
//! 0.5*(target[fl-1] + target[fl]) computed as (a+b)*0.5.

/// Quantile-normalise columns of `y` (row-major, rows x cols).
pub fn quantile_normalize(y: &[f64], rows: usize, cols: usize) -> Vec<f64> {
    // 1) target: row_mean[i] += sorted_col_j[i] / cols, column by column
    let inv_cols = cols as f64;
    let mut row_mean = vec![0.0f64; rows];
    let mut col_buf = vec![0.0f64; rows];
    for j in 0..cols {
        for i in 0..rows {
            col_buf[i] = y[i * cols + j];
        }
        col_buf.sort_by(|a, b| a.partial_cmp(b).unwrap());
        for i in 0..rows {
            row_mean[i] += col_buf[i] / inv_cols;
        }
    }

    // 2) distribute with averaged ranks for ties (get_ranks: (i+k+2)*0.5)
    let mut out = vec![0.0f64; rows * cols];
    let mut order: Vec<usize> = (0..rows).collect();
    let mut ranks = vec![0.0f64; rows];
    for j in 0..cols {
        for i in 0..rows {
            col_buf[i] = y[i * cols + j];
        }
        order.sort_by(|&a, &b| col_buf[a].partial_cmp(&col_buf[b]).unwrap());
        let sv: Vec<f64> = order.iter().map(|&i| col_buf[i]).collect();
        let mut i = 0;
        while i < rows {
            let mut k = i;
            while k < rows - 1 && sv[k] == sv[k + 1] {
                k += 1;
            }
            if k != i {
                let r = (i + k + 2) as f64 * 0.5;
                for t in i..=k {
                    ranks[t] = r;
                }
            } else {
                ranks[i] = (i + 1) as f64;
            }
            i = k + 1;
        }
        for t in 0..rows {
            let fl = ranks[t].floor() as usize; // 1-based floor
            let frac = ranks[t] - fl as f64;
            let lo = row_mean[fl - 1];
            let val = if frac > 0.4 {
                (lo + row_mean[fl.min(rows - 1)]) * 0.5
            } else {
                lo
            };
            out[order[t] * cols + j] = val;
        }
    }
    out
}
